chain/bitcoind_events.go

```diff
@@ -0,0 +1,321 @@
+		// that it succeeded but the tx zmq connection failed.
+		if err := zmqBlockConn.Close(); err != nil {
+			log.Errorf("could not close zmq block conn: %v", err)
+		}
+
+		return nil, fmt.Errorf("unable to subscribe for zmq tx "+
+			"events: %v", err)
+	}
+
+	return &bitcoindZMQEvents{
+		cfg:         cfg,
+		blockConn:   zmqBlockConn,
+		txConn:      zmqTxConn,
+		notifyBlock: notifyBlock,
+		notifyTx:    notifyTx,
+		quit:        make(chan struct{}),
+	}, nil
+}
+
+// Start spins off the bitcoindZMQEvent goroutines.
+func (b *bitcoindZMQEvents) Start() error {
+	b.wg.Add(2)
+	go b.blockEventHandler()
+	go b.txEventHandler()
+	return nil
+}
+
+// blockEventHandler reads raw blocks events from the ZMQ block socket and
+// forwards them along to the current rescan clients.
+//
+// NOTE: This must be run as a goroutine.
+func (b *bitcoindZMQEvents) blockEventHandler() {
+	defer b.wg.Done()
+
+	log.Info("Started listening for bitcoind block notifications via ZMQ "+
+		"on", b.blockConn.RemoteAddr())
+
+	// Set up the buffers we expect our messages to consume. ZMQ
+	// messages from bitcoind include three parts: the command, the
+	// data, and the sequence number.
+	//
+	// We'll allocate a fixed data slice that we'll reuse when reading
+	// blocks from bitcoind through ZMQ. There's no need to recycle this
+	// slice (zero out) after using it, as further reads will overwrite the
+	// slice and we'll only be deserializing the bytes needed.
+	var (
+		command [len(rawBlockZMQCommand)]byte
+		seqNum  [seqNumLen]byte
+		data    = make([]byte, maxRawBlockSize)
+	)
+
+	for {
+		// Before attempting to read from the ZMQ socket, we'll make
+		// sure to check if we've been requested to shut down.
+		select {
+		case <-b.quit:
+			return
+		default:
+		}
+
+		// Poll an event from the ZMQ socket.
+		var (
+			bufs = [][]byte{command[:], data, seqNum[:]}
+			err  error
+		)
+		bufs, err = b.blockConn.Receive(bufs)
+		if err != nil {
+			// EOF should only be returned if the connection was
+			// explicitly closed, so we can exit at this point.
+			if err == io.EOF {
+				return
+			}
+
+			// It's possible that the connection to the socket
+			// continuously times out, so we'll prevent logging this
+			// error to prevent spamming the logs.
+			netErr, ok := err.(net.Error)
+			if ok && netErr.Timeout() {
+				log.Trace("Re-establishing timed out ZMQ " +
+					"block connection")
+				continue
+			}
+
+			log.Errorf("Unable to receive ZMQ %v message: %v",
+				rawBlockZMQCommand, err)
+			continue
+		}
+
+		// We have an event! We'll now ensure it is a block event,
+		// deserialize it, and report it to the different rescan
+		// clients.
+		eventType := string(bufs[0])
+		switch eventType {
+		case rawBlockZMQCommand:
+			block := &wire.MsgBlock{}
+			r := bytes.NewReader(bufs[1])
+			if err := block.Deserialize(r); err != nil {
+				log.Errorf("Unable to deserialize block: %v",
+					err)
+				continue
+			}
+
+			b.notifyBlock(block)
```

Comment:
So with my suggestion above, this here would instead select on the channel write or quit.